RIOT-OS/RIOT
# treewide: Use deprecations in code rather than comments

This is a replacement PR for #18565  because that had been made to the wrong branch and GitHub closed it irrevertably when I tried to clean up.

The relevant parts of that PR are replicated here:

---

### Contribution description

This executes <https://github.com/RIOT-OS/RIOT/issues/18561>.

It already has the macros and Doxygen stuff, it lacks
* change of all `@deprecated` to `DEPRECATED`
* CI time rejection of non-code deprecation (unsure if possible, maybe we still need it for module level deprecations)
* Removal of deprecated functions in the code base, or annotation with `-Wno-error=deprecated-declarations` if we still want to test the deprecated functions.

### Testing procedure

Right now:

* See how CI starts breaking because we -Werror and the one function that got changed is still used in tests.
* Look at Doxygen output on the changed function.
* [edit: added] Also try building an example of a harder deprecation with a newer C standard where the actual C version (and not the GCC proprietary one) is used:
```
$ CFLAGS=-std=c2x make -C examples/usbus_minimal BOARD=samr21-xpro
```

[edit: or the more forgiving gnu2x, given that native has some trouble with being precisely conformant]

Long term:

* Try using a deprecated function, see things fail.

---

@maribu said:

If you throw in an `-Wno-error=deprecated-declarations`, the warning should not be escalated to an error. We may however want to still escalate it to an error with `RIOT_CI_BUILD=1` to avoid in-tree users sneaking back in.

--

@chrysn responded:

On the severity of errors, I think a good path would be to have Makefiles of `ALLOW_DEPRECATED ?= 1`, make Murdock set `ALLOW_DEPRECATED = 0`, and have some tests (those that explicitly include deprecated functions because we don't want these to accidentally break) set `ALLOW_DEPRECATED = 1`.

---

On the CI side, Murdock went well because it duly rejected the tests that used the deprecated function (even though, as above, these should later be exempt).

---

Sadly, doxygen output [breaks completely](https://output.circle-artifacts.com/output/job/e2b7691b-ea06-4bbb-a7ed-f16e0c77353f/artifacts/0/doc/group__drivers__periph__flashpage.html) with the current doxygen version -- deprecation text gets moved into functions list, but the function itself is not present any more at all, and the next function is somehow also missing ... putting this in as "waiting for CI update" because the next Ubuntu hopefully has a better Doxygen. (Things looked good with the 1.9.4 I have here).

This is now blocked by https://github.com/RIOT-OS/riotdocker/pull/189
